GKE GPU A3 with TCPX support

samos123 opened the pull request:
Add the ability to launch GPU jobs directly on GKE

**Test cases I ran**:

Fuji 7b single host per device batch 8 is working too:
```
BASTION_TIER=1 axlearn gcp gke start --instance_type=gpu-a3-highgpu-8g-8 \
        --cluster=daily-run-single-zone-gker27 --num_replicas=1 \
        --queue multislice-queue \
        --bundler_spec=allow_dirty=True \
        --bundler_type=artifactregistry --bundler_spec=image=gpu \
        --bundler_spec=dockerfile=Dockerfile --bundler_spec=target=gpu \
        -- mkdir -p /tmp/test_trainer\; python3 -m axlearn.common.launch_trainer_main \
  --module=text.gpt.c4_trainer --config=fuji-7B-v1-single-host \
  --trainer_dir=/tmp/test_trainer --data_dir=gs://axlearn-public/tensorflow_datasets \
  --jax_backend=gpu
```
step time on single node: 2.06 seconds

Fuji 7b 32 nodes batch per device 8:
```
BASTION_TIER=1 axlearn gcp gke start --instance_type=gpu-a3-highgpu-8g-256 \
        --cluster=daily-run-single-zone-gker27 --num_replicas=32 \
        --queue multislice-queue \
        --bundler_spec=allow_dirty=True \
        --bundler_type=artifactregistry --bundler_spec=image=gpu \
        --bundler_spec=dockerfile=Dockerfile --bundler_spec=target=gpu \
        -- mkdir -p /tmp/test_trainer\; python3 -m axlearn.common.launch_trainer_main \
  --module=text.gpt.c4_trainer --config=fuji-7B-v1 \
  --trainer_dir=/tmp/test_trainer --data_dir=gs://axlearn-public/tensorflow_datasets \
  --jax_backend=gpu
```
step time on 32 nodes: 3.03 seconds

Fuji 7b 64 nodes batch per device 4:
```
BASTION_TIER=1 axlearn gcp gke start --instance_type=gpu-a3-highgpu-8g-512 \
        --cluster=daily-run-single-zone-gker27 --num_replicas=64 \
        --queue multislice-queue \
        --bundler_spec=allow_dirty=True \
        --bundler_type=artifactregistry --bundler_spec=image=gpu \
        --bundler_spec=dockerfile=Dockerfile --bundler_spec=target=gpu \
        -- mkdir -p /tmp/test_trainer\; python3 -m axlearn.common.launch_trainer_main \
  --module=text.gpt.c4_trainer --config=fuji-7B-v1 \
  --trainer_dir=/tmp/test_trainer --data_dir=gs://axlearn-public/tensorflow_datasets \
  --jax_backend=gpu
```
Step time A3 3.92 seconds (note A3+ this is 1.785 seconds).

I plan to work on A3+ support in a separate PR.

samos123 commented on axlearn/cloud/gcp/job.py:
```diff
@@ -540,9 +540,9 @@ def _build_pod(self) -> Nested[Any]:
-                if cfg.enable_pre_provisioner
-                else "false",
+                "tpu-provisioner.cloud.google.com/disable-autoprovisioning": (
+                    "true" if cfg.enable_pre_provisioner else "false"
```
this is due to black formatter running I think

samos123 replied:
closing since you seem to be fine with this

markblee commented on Dockerfile:
```diff
@@ -90,6 +90,25 @@ ENV PIP_FIND_LINKS=https://storage.googleapis.com/jax-releases/libtpu_releases.h
+
+RUN apt-get install -y google-perftools
+
+RUN JAX_VERSION=$(grep "jax==" pyproject.toml | sed -E 's/"jax==([0-9.]+)",/\1/') && \
+    pip install --no-cache-dir --upgrade \
+      "jax[cuda12_pip]==${JAX_VERSION}" \
+      -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html
```
Could this be in pyproject too, since we already specify `PIP_FIND_LINKS` below?

samos123 replied:
fixed

markblee commented on axlearn/cloud/gcp/job.py:
```diff
@@ -639,6 +639,327 @@ def _execute(self) -> Any:
+        cfg: GPUGKEJob.Config = self.config
+
+        if cfg.accelerator.instance_type.startswith("a3-highgpu"):
+            volume_mounts = [
```
Add a reference/comment on what this sidecar is doing and why we only do this for a3-highgpu?

samos123 replied:
done

markblee commented on axlearn/cloud/gcp/job.py:
```diff
@@ -639,6 +639,327 @@ def _execute(self) -> Any:
+        """
+        cfg: GPUGKEJob.Config = self.config
+        volume_mounts = []
+        if cfg.accelerator.instance_type.startswith("a3-highgpu"):
```
nit suggestion: construct the default volume mounts, env vars, etc. upfront, and then group all of the `a3-highgpu` specific changes under one branch.

samos123 replied:
the only volume mount that's needed by any machine type would likely be `{"name": "shared-memory", "mountPath": "/dev/shm"},` so I've moved that one out of the if

samos123 replied:
I also changed to a single if statement for a3-highpu specific changes. Marking as fixed

markblee commented on axlearn/cloud/gcp/job.py:
```diff
@@ -639,6 +639,327 @@ def _execute(self) -> Any:
+        env_vars["DISTRIBUTED_COORDINATOR"] = f"{cfg.name}-job-0-0.{cfg.name}:8080"
+        env_vars["NUM_PROCESSES"] = f"{cfg.accelerator.num_replicas}"
+        if cfg.accelerator.instance_type.startswith("a3-highgpu"):
+            env_vars["LD_LIBRARY_PATH"] = "/usr/local/tcpx/lib64:/usr/local/nvidia/lib64"
```
nit -- `env_vars.update({ ... })` may be slightly easier to read.

samos123 replied:
fixed

markblee commented on axlearn/cloud/gcp/job.py:
```diff
@@ -639,6 +639,327 @@ def _execute(self) -> Any:
+
+        command = ["bash", "-c", cfg.command]
+        if cfg.accelerator.instance_type.startswith("a3-highgpu"):
+            command.append("touch /run/tcpx/terminated")
```
What is this file used for? Is it read by the sidecar?

samos123 replied:
I will add a comment to the code, but this is required so the sidecar exits once the training job is done, otherwise the pod will just stay running forever

samos123 replied:
It's basically a way to signal to the sidecar container to exit

markblee commented on axlearn/common/launch.py:
```diff
@@ -70,7 +70,7 @@
 flags.DEFINE_string(
     "distributed_coordinator",
-    None,
+    os.environ.get("DISTRIBUTED_COORDINATOR", None),
```
Do we need this change for the GPU runner, or can we just supply the flags for now?

samos123 replied:
It gets ugly imo if you have to pass through these flags 2 times. You will end up having to pass it through as environment variables to the pod and then read those env variables within the container to pass it as flags.

Wouldn't it be nicer if we can pass env variables as default instead and only pass flags to override if needed?

The way it's implemented is backwards compatible, it will still return None as the default if the env flag is not set.

samos123 replied:
For example, without pulling from env variables I would have to do this:
```
BASTION_TIER=1 axlearn gcp gke start --instance_type=gpu-a3-highgpu-8g-8 \
        --cluster=daily-run-single-zone-gker27 --num_replicas=1 \
        --queue multislice-queue \
        --bundler_type=artifactregistry --bundler_spec=image=gpu \
        --bundler_spec=dockerfile=Dockerfile --bundler_spec=target=gpu \
        -- mkdir -p /tmp/test_trainer\; python3 -m axlearn.common.launch_trainer_main \
  --module=text.gpt.c4_trainer --config=fuji-7B-v1-single-host \
  --trainer_dir=/tmp/test_trainer --data_dir=gs://axlearn-public/tensorflow_datasets \
  --jax_backend=gpu --distributed_coordinator=\$DISTRIBUTED_COORDINATOR \
  --process_id=\$PROCESS_ID ---num_processes=\$NUM_PROCESSES
  ```

markblee commented on axlearn/cloud/gcp/jobs/gke_runner.py:
```diff
@@ -446,9 +446,18 @@ def from_flags(cls, fv: flags.FlagValues, **kwargs):
+    """A GKERunnerJob that uses GPUGKEJob."""
+
+    inner = GPUGKEJob
+    pre_provisioner = TPUNodePoolProvisioner
```
A reminder to change `TPUNodePoolProvisioner` when ready. Maybe we can just use the default implementation that raises NotImplementedError.

samos123 replied:
I somehow missed this! Thanks for catching it. Will fix

samos123 replied:
fixed

markblee commented on axlearn/cloud/gcp/job.py:
```diff
@@ -639,6 +639,327 @@ def _execute(self) -> Any:
+    def _build_init_container(self) -> Nested[Any]:
+        """Builds a config for a single container."""
+        cfg: GPUGKEJob.Config = self.config
+        if cfg.accelerator.instance_type.startswith("a3-highgpu"):
```
I noticed that we have this check in a lot of places. What are the other instance types that we intend to support?

samos123 replied:
I am prepping to also add a3-mega which will require it's own unique setup and configuration sadly. I will think of a way to make it cleaner but we do need machine type specific checks for A3/A3 Mega/ A3 Ultra/ A4

samos123 replied:
I've removed this check for now and will add support for multiple instance type e.g. a3 MEGA at a later time

markblee reviewed (commented):
A quick skim through...

samos123 commented on axlearn/experiments/text/gpt/c4_trainer.py:
```diff
@@ -109,6 +109,9 @@ def named_trainer_configs() -> Dict[str, TrainerConfigFn]:
             max_sequence_length = kwargs.pop("max_sequence_length")
+
+            # TODO remove before merging
+            kwargs["max_step"] = 1000
```
how can I do this at runtime? It makes it harder to experiment when the max step is so large. I want to be able to pass this through as a command line parameter.

markblee replied:
You can create a separate module with your changes and point to it via `--module`, similar to your previous fuji experiments. Let me know if you prefer a more concrete example.

samos123 replied:
yeah I know how to do it. I still don't like it. I prefer to be able to specify max_step through CLI so I can use the upstream modules for testing my PRs. I do get wanting to express everything as code so feel free to ignore this.

markblee replied:
While `max_step` is probably a simple case, often times trainer configs have multiple layers of composition which is not really conducive to flag overrides. Code is also more traceable/reviewable, as experiments can be tracked in commit history (and golden configs) rather than someone's bash history.

markblee commented on axlearn/cloud/gcp/job.py:
```diff
@@ -639,6 +639,359 @@ def _execute(self) -> Any:
+                ]
+            )
+            env_vars.update(
+                {
```
I think this whole section can benefit from more comments/pointers to what these are doing or how we decided on the defaults.

samos123 replied:
I will give it a try since these are the required NCCL env variables provided by our eng team. I will try to add some comments and see if I can get some feedback from our eng team.

markblee reviewed (commented):
Thanks, overall looks reasonable. Can we add/update tests for changes in `job.py` and `gke_runner.py`?

samos123 commented on axlearn/cloud/gcp/job.py:
```diff
@@ -639,6 +639,359 @@ def _execute(self) -> Any:
+        """Configures GPUGKEJob.
+
+        Attributes:
+            accelerator: GPU configuration.
```
fixed

samos123 commented on axlearn/cloud/gcp/job.py:
```diff
@@ -639,6 +639,359 @@ def _execute(self) -> Any:
+        default_xla_flags = [
+            "--xla_gpu_enable_latency_hiding_scheduler=true",
+            "--xla_gpu_all_reduce_contiguous",
+            "--xla_gpu_all_reduce_combine_threshold_bytes=1073741824",
+            "--xla_gpu_all_gather_combine_threshold_bytes=1073741824",
+            "--xla_gpu_reduce_scatter_combine_threshold_bytes=1073741824",
```
I think these are good defaults that bring better performance for all or almost all workloads. The workloads can still override it by providing:
```
axlearn gke start ... --env 'XLA_FLAGS: "--xla_gpu_enable_latency_hiding_scheduler=false"'
```

The XLA defaults perform significantly worse.

So I would prefer to keep these as-is to so everyone gets a decent out of the box performance with GCP GPUs.

samos123 replied:
fixed

samos123 commented on axlearn/cloud/gcp/job.py:
```diff
@@ -639,6 +639,359 @@ def _execute(self) -> Any:
+        user_cmd = cfg.command
+        # This is needed to make the sidecar exit when the main container exits.
+        if cfg.accelerator.instance_type.startswith("gpu-a3-highgpu"):
+            user_cmd += "\ntouch /run/tcpx/terminated"
```
Yes it's expected. New line works like a semicolon in this case. I verified this is working. Let me know if you prefer a semicolon though, because that should work too. I can retest with a semicolon if that's your preference.

samos123 replied:
fixed

samos123 commented on axlearn/cloud/gcp/job.py:
```diff
@@ -639,6 +639,359 @@ def _execute(self) -> Any:
+        cfg: GPUGKEJob.Config = self.config
+
+        # Different machine types require different sidecar containers
+        # for example A3 requires a tcpx socket but A3 Mega does not
```
fixed

samos123 commented:
I have addressed all comments, some with fixes and some with responses instead. 

I've added tests for GPUGKEJob. However, the GPUGKERunner is a single line wrapper so testing coverage doesn't seem helpful there. I would prefer skipping adding tests to gke_runner_test.py. Otherwise some guidance on what GPUGKERunner scenario you would like to see me test would be helpful.

samos123 commented on axlearn/cloud/gcp/job.py:
```diff
@@ -639,6 +639,384 @@ def _execute(self) -> Any:
+        instance_type = cfg.accelerator.instance_type
+        if not instance_type.startswith("gpu-a3-highgpu"):
+            raise NotImplementedError(
+                f"The instance type {instance_type} is not supported on GKE with GPU."
+                "Only gpu-a3-highgpu-8g is supported"
```
I do plan to have a follow up PR for a3-megagpu very soon. That being said, to keep code simple for today's needs, it may be better to remove all the if checks and come up with a better mechanism once we support multiple types.

samos123 replied:
fixed

markblee replied:
Should we fix the formatting?

samos123 replied:
pushed and closing

samos123 commented on axlearn/cloud/gcp/job.py:
```diff
@@ -639,6 +639,384 @@ def _execute(self) -> Any:
+                "valueFrom": {
+                    "fieldRef": {
+                        "fieldPath": (
+                            "metadata.annotations['batch.kubernetes.io/" "job-completion-index']"
```
fixed

samos123 commented:
I did another test on single node but still waiting for multi-node results. The single node result was able to verify that `;` works instead of `\n`.

markblee commented on axlearn/cloud/gcp/job.py:
```diff
@@ -639,6 +639,384 @@ def _execute(self) -> Any:
+
+        user_cmd = cfg.command
+        if user_cmd is None:
+            user_cmd = ""
```
Was the change pushed?

samos123 replied:
yes, marking as fixed

markblee reviewed (commented):
Can we leave a comment "Done" for comments which have been addressed? (See also: https://github.com/apple/axlearn/blob/main/CONTRIBUTING.md)

samos123 commented:
I resolved all comments and fixed missing punctation

markblee reviewed (approved):
Thanks!

samos123 commented:
Seems Circle CI is awaiting someone to approve the run?

samos123 commented:
Seems it was approved but there was a Circe CI Apple Auth error: https://app.circleci.com/pipelines/github/apple/axlearn/1925/workflows/99dbcc0a-7a9f-4338-8ec3-0cc2724a9e28
I can't see what the error is though. So will probably need @markblee to take a look?

samos123 commented:
Looks like Apple Authenticate in circle ci is still failing but I don't see the logs.
